File: src/R/src/estimators.r
library(assertthat)

NO_BEST <-c(0,0)

Leaf <- function(value,id=0,bool_id=NULL){
  assert_that(is.numeric(value))
  assert_integer(id)
  assert_that(is.logical(bool_id)||is.null(bool_id))
  
  id <-ifelse(id==0&&!is.null(bool_id),bool_vector_to_int(bool_id),id)
  leaf <- list(id=id,value=value)
  attr(leaf, "class") <- "leaf"
  leaf
}

length.leaf <- function(obj){1}
depth.leaf <- function(obj){0}

Node <- function(node_bool_id,feat_id,featval,left,right){
  assert_that(is.logical(node_bool_id))
  assert_integer(feat_id)
  assert_that(is(left,"leaf")||is(left,"node")||is.null(left))
  assert_that(is(right,"leaf")||is(right,"node")||is.null(right))
  
  node <- list(id=bool_vector_to_int(node_bool_id),feat_id = feat_id,
               featval = featval,left=left,right = right)
  attr(node, "class") <- "node"
  node
}

length.node <- function(obj){
  length(obj$left)+length(obj$right)
}

depth.node <-function(obj){
  1 + max(depth(obj$left), depth(obj$right))
}

fit_regression_tree <- function(X,Y,min_samples_leaf=5,max_depth=-1, node_id_bool=TRUE){
  #May be needed to remove the is.vector assert
  assert_that((is.matrix(X)||is.vector(X))&&is.numeric(X))
  assert_that(is.vector(Y)&&is.numeric(Y))
  assert_integer(min_samples_leaf)
  assert_integer(max_depth)
  assert_integer(min_samples_leaf)
  assert_that(is.logical(node_id_bool))
  
  if (max_depth < -1){
    stop(cat("Unexpected value for max_depth: ",max_depth,
             " (expected: max_depth >= 0, or max_depth = -1 for infinite depth)"))
  }
  
  if (length(Y) <= min_samples_leaf || max_depth==0){
     Leaf(mean(Y),bool_id = node_id_bool)
  }
  else{
    S <- split_mse(X, Y)
    
    if(identical(S,NO_BEST)){
      Leaf(mean(Y),bool_id = node_id_bool)
    }
    else{
      feat_id <- S[1];thresh = S[2]
      
      split <- X[,feat_id]< thresh
      
      max_depth <- max(max_depth-1, -1)
      
      Node(node_id_bool, feat_id, thresh,
           
                  fit_regression_tree(X[split,], Y[split],min_samples_leaf, max_depth, 
                                      node_id_bool =c(TRUE,node_id_bool)),
           
                  fit_regression_tree(X[!split,],Y[!split],min_samples_leaf,max_depth, 
                                      node_id_bool =c(FALSE,node_id_bool))
      )
    }
  }
}
  

split_mse <- function(X,Y){
  
  assert_that(is.matrix(X)&&is.numeric(X))
  assert_that(is.vector(Y)&&is.numeric(Y))
  
  N<-length(X[,1])
  p<-length(X[1,])
  
  best <- NO_BEST
  best_val <- -Inf
  
  for (i in 1:p){
    ord <- order(X[,i])
    X_i <- X[ord,i]
    Y_i <- Y[ord]
    
    if(N > 100){
      domain_i <- quantile(X_i,seq(0.01,1,by=0.01))
    }
    else{
      domain_i = X_i
    }
    
    mse<-best_mse_loss(X_i, Y_i, domain_i)
    
    value <- mse[1]; thresh <- mse[2] 
    if (value > best_val){
      best_val <- value
      best<-c(i,thresh)
      #mse is zero and all Y are equal -> no split 
      if(isTRUE(all.equal(best_val,0))){
        if(are_elements_equal(Y_i)){
          best<-NO_BEST
        }
      }
    }
  }
  best
}

best_mse_loss<-function(X,Y,domain){
  assert_that(is.vector(X)&&is.numeric(X))
  assert_that(is.vector(Y)&&is.numeric(Y))
  assert_that(is.vector(domain)&&is.numeric(domain))
  #assert_that(!is.unsorted(X))
  
  best_val <- -Inf
  best_thresh <- 0
  
  s_l <-0; s2_l = 0
  
  su <- sum(Y)
  su2<-0
  
  for (l in Y){
    su2 <- su2 + l*l
  } 
  
  nl<-0
  n<-length(Y)
  i<-1
  
  for (thresh in domain){
    while (i <= length(Y) && X[i] < thresh){
      l <- Y[i]
      
      s_l <- s_l + l
      s2_l <- s2_l + l*l
      nl <- nl + 1
      
      i <- i + 1
    }

    s_r <- su - s_l
    s2_r <- su2 - s2_l
    nr <- n - nl

    if (nr > 0 && nl > 0){
      
      loss <- s2_l - (s_l^2)/nl + s2_r - (s_r^2)/nr
      
      if (-loss > best_val){
        best_val <- (-loss)
        best_thresh <- thresh
      }
    }
  }
  c(best_val, best_thresh)
}


bool_vector_to_int <- function(bool_vect){
  assert_that(is.logical(bool_vect))
  
  res <- 0
  counter <- 0
  for (v in bool_vect){
    res= res + (2^counter)*v
    counter = counter+1
  } 
  res
}

depth <- function(obj) {
  UseMethod("depth")
}

predict_treated_dictionary <- function(estimator_pair,X,assign=FALSE){
  assert_that(is(estimator_pair,"treatment_dictionary"))
  assert_that(is.matrix(X))
  assert_that(is.logical(assign))
  
  predictions<-NumericTreatmentDictionary(0,0)
  for(tr in names(predictions)){
    predictions[[tr]]<-predict_rt_matrix(estimator_pair[[tr]],X,assign)
  }
  predictions
}

predict_rt_matrix <- function(tree,X,assign=FALSE){
  assert_that(is(tree,"leaf")||is(tree,"node"))
  assert_that(is.matrix(X))
  assert_that(is.logical(assign))
  
  N = length(X[,1])
  predictions <- rep(0,times=N)
  for(i in 1:N){
    predictions[i]<-predict_rt(tree,X[i,],assign)
  }
  predictions
} 

predict_rt <- function(tree,X,assign=FALSE){
  assert_that(is(tree,"leaf")||is(tree,"node"))
  assert_that(is.vector(X))
  assert_that(is.logical(assign))
  
  if(is(tree,"leaf")){
    if(assign){
      tree$id
    }
    else{
      tree$value 
    }
  }
  else{
    if(is.null(tree$featval)){
       predict_rt(tree$left, X,assign)
    }
    else if(X[tree$feat_id]<tree$featval){
      predict_rt(tree$left, X,assign)
    }
    else{
      predict_rt(tree$right, X,assign)
    }
  }
}

depth.default <- function(obj){0}

File: src/Julia/estimators.jl
# Adapted from https://github.com/bensadeghi/DecisionTree.jl
__precompile__()

module Estimators
import Base: length, convert, promote_rule, show, start, next, done
export Estimator, Leaf, Node,
       depth,
       fit_regression_tree,
       predict, assign_leaves

float(x) = map(Float64, x)
neg(arr) = map(!, arr) # `.!arr` is invalid in 0.5, and `!arr` triggers a warning in 0.6.
const NO_BEST=(0,0)

# converts a boolean vector into an integer by summing the appropriate powers of 2
boolvec2int(v::Vector{Bool}) = sum(b*v for (b,v) in zip(2.^(0:(length(v)-1)), v))

abstract type Estimator
end

immutable Leaf <: Estimator
    id::Integer
    value::Real
end
    length(leaf::Leaf) = 1
    depth(leaf::Leaf) = 0

Leaf(id::Vector{Bool}, value::Real) = Leaf(boolvec2int(id), value)
Leaf(value::Real) = Leaf(0,value)

immutable Node <: Estimator
    node_id::Integer
    featid::Integer
    featval::Any
    left::Union{Leaf,Node}
    right::Union{Leaf,Node}
end
    length(tree::Node) = length(tree.left) + length(tree.right)
    depth(tree::Node) = 1 + max(depth(tree.left), depth(tree.right))

function Node(id::Vector{Bool}, featid::Integer, featval::Any, left::Union{Leaf,Node}, right::Union{Leaf,Node})
    return Node(boolvec2int(id), featid, featval, left, right)
end

convert(::Type{Node}, x::Leaf) = Node(0, nothing, x, Leaf(nothing,[nothing])) #makes a node with that leaf as it's left node
promote_rule(::Type{Node}, ::Type{Leaf}) = Node # this tells julia to call convert(::node, leaf) on the leaf when this happens
promote_rule(::Type{Leaf}, ::Type{Node}) = Node

function fit_regression_tree{T<:Float64, U<:Real}(X::Matrix{U}, Y::Vector{T}; min_samples_leaf=5, max_depth=-1, node_id::Vector{Bool}=[true])
    if max_depth < -1
        error("Unexpected value for max_depth: $(max_depth) (expected: max_depth >= 0, or max_depth = -1 for infinite depth)")
    end
    if length(Y) <= min_samples_leaf || max_depth==0
        return Leaf(node_id, mean(Y))

    end
    S = _split_mse(X, Y)
    if S == NO_BEST
        return Leaf(node_id, mean(Y))
    end
    feat_id, thresh = S
    split = X[:,feat_id] .< thresh
    return Node(node_id, feat_id, thresh,
                fit_regression_tree(X[split,:], Y[split],
                    min_samples_leaf=min_samples_leaf, max_depth=max(max_depth-1, -1), node_id=vcat(true,node_id)),
                fit_regression_tree(X[neg(split),:], Y[neg(split)],
                    min_samples_leaf=min_samples_leaf, max_depth=max(max_depth-1, -1), node_id=vcat(false,node_id))
                )
end

function _split_mse{T<:Float64, U<:Real}(X::Matrix{U}, Y::Vector{T})
    N, p = size(X)

    best = NO_BEST
    best_val = -Inf

    for i in 1:p
        ord = sortperm(X[:,i])
        X_i = X[ord,i]
        Y_i = Y[ord]
        if N > 100
            if VERSION >= v"0.4.0-dev"
                domain_i = quantile(X_i, linspace(0.01, 0.99, 99);
                                    sorted=true)
            else  # sorted=true isn't supported on StatsBase's Julia 0.3 version
                domain_i = quantile(X_i, linspace(0.01, 0.99, 99))
            end
        else
            domain_i = X_i
        end
        value, thresh = _best_mse_loss(X_i, Y_i, domain_i)
        if value > best_val
            best_val = value
            best = (i, thresh)
        end
    end

    return best
end


function _best_mse_loss{T<:Float64, U<:Real}(X::Vector{U}, Y::Vector{T}, domain)
    # True, but costly assert. However, see
    # https://github.com/JuliaStats/StatsBase.jl/issues/164
    # @assert issorted(X) && issorted(domain)
    best_val = -Inf
    best_thresh = 0.0
    s_l = s2_l = zero(T)
    su = sum(Y)::T
    su2 = zero(T); for l in Y su2 += l*l end  # sum of squares
    nl = 0
    n = length(Y)
    i = 1
    # Because the `X` are sorted, below is an O(N) algorithm for finding
    # the optimal threshold amongst `domain`. We simply iterate through the
    # array and update s_l and s_r (= sum(Y) - s_l) as we go. - @cstjean
    @inbounds for thresh in domain
        while i <= length(Y) && X[i] < thresh
            l = Y[i]

            s_l += l
            s2_l += l*l
            nl += 1

            i += 1
        end
        s_r = su - s_l
        s2_r = su2 - s2_l
        nr = n - nl
        # This check is necessary I think because in theory all Y could
        # be the same, then either nl or nr would be 0. - @cstjean
        if nr > 0 && nl > 0
            loss = s2_l - s_l^2/nl + s2_r - s_r^2/nr
            if -loss > best_val
                best_val = -loss
                best_thresh = thresh
            end
        end
    end
    return best_val, best_thresh
end

predict(leaf::Leaf, feature::Vector) = leaf.value

function predict(tree::Node, X::Vector)
    if tree.featval == nothing
        return predict(tree.left, X)
    elseif X[tree.featid] < tree.featval
        return predict(tree.left, X)
    else
        return predict(tree.right, X)
    end
end

function predict(tree::Union{Leaf,Node}, X::Matrix)
    N = size(X,1)
    predictions = Array{Any}(N)
    for i in 1:N
        predictions[i] = predict(tree, X[i,:])
    end
    println("predictions ->")
    println(predictions)
    if typeof(predictions[1]) <: Float64
        return float(predictions)
    else
        return predictions
    end
end

function predict(estimator_dict::Dict, X::Matrix)
    return Dict(k=>predict(v,X) for (k,v) in estimator_dict)
end

assign_leaves(leaf::Leaf, feature::Vector) = leaf.id

function assign_leaves(tree::Node, X::Vector)
    if tree.featval == nothing
        return assign_leaves(tree.left, X)
    elseif X[tree.featid] < tree.featval
        return assign_leaves(tree.left, X)
    else
        return assign_leaves(tree.right, X)
    end
end

function assign_leaves(tree::Union{Leaf,Node}, X::Matrix)
    N = size(X,1)
    assignments = Array{Int}(N)
    for i in 1:N
        assignments[i] = assign_leaves(tree, X[i,:])
    end
    return assignments
end

function assign_leaves(estimator_dict::Dict, X::Matrix)
    return Dict(k=>assign_leaves(v,X) for (k,v) in estimator_dict)
end

end #module
